### PR TITLE
Using Solution builder to build project references

### PR DESCRIPTION
**Pull request by sheetalkamat**
This is based off microsoft/TypeScript#31432

### CODE REVIEW

**Comment by johnnyreilly** [comment]
Very exciting! Looks like there's a compilation error at the moment - I'm guessing we're just waiting for the relevant API to appear in `@next` 😁

**Comment by sheetalkamat** [comment]
@johnnyreilly Yes. The PR is almost ready for review and once the build should be able to succeed.

**Comment by johnnyreilly** [comment]
Cool - let me know when you're ready and we've a passing build and I'll take a look.

I'm guessing the new execution test you've added should pass with the `typescript@next` test run.

**Comment by sheetalkamat** [comment]
@andrewbranch Can you please try it out now. SolutionBuilder api is in typescript nightly now and I have updated this PR.

**Comment by sheetalkamat** [comment]
@johnnyreilly This is ready to go in. Please take a look. Thanks.

**Comment by johnnyreilly** [comment]
Thanks @sheetalkamat! If I get time I'll try to look this weekend. Am I right in thinking the SolutionBuilder api will be released with TypeScript 3.6?

**Comment by sheetalkamat** [comment]
@johnnyreilly Interesting not sure what I can do to fix that, but I can run that as a single test and it passes.

**Comment by andrewbranch** [comment]
I was blocked by https://github.com/microsoft/TypeScript/issues/31792 trying to test this in the wild yesterday, but should be able to play with it today.

**Comment by johnnyreilly** [comment]
It probably makes more sense for @andrewbranch to carry on with the reviewing / merging of this. He's much more knowledgeable about project references than I am.

If he's happy then I'm happy 😀

File: src/servicesHost.ts [hunk 1]
```diff
@@ -670,9 +749,7 @@ function addCache(servicesHost: typescript.ModuleResolutionHost) {
       const cache = createCache<ReturnType<typeof originalFunction>>(
         originalFunction
       );
-      servicesHost[
-        functionToCache
-      ] = cache.getCached as typescript.ModuleResolutionHost[CacheableFunction];
+      servicesHost[functionToCache] = cache.getCached as any;
```
Comment (johnnyreilly) [inline src/servicesHost.ts]: could we fix the `any` please?

Comment (johnnyreilly) [reply src/servicesHost.ts]: See also https://github.com/TypeStrong/ts-loader/pull/954#discussion_r294036783 where @fa93hws has bumped on this too.

**Comment by artem-malko** [comment]
Hi all)
Any news?

File: src/config.ts [hunk 2]
```diff
@@ -149,3 +150,31 @@ export function getConfigParseResult(
...
+    loaderOptions.compilerOptions,
+    {
+      ...compiler.sys,
+      onUnRecoverableConfigFileDiagnostic: () => {} // tslint:disable-line no-empty
+    },
+    extendedConfigCache
+  );
+  if (result) {
+    const options = getCompilerOptions(result);
+    (compiler as any).setConfigFileInOptions(
```
Comment (johnnyreilly) [inline src/config.ts]: Is the `any` necessary?

File: src/config.ts [hunk 3]
```diff
@@ -149,3 +150,31 @@ export function getConfigParseResult(
...
+      options,
+      (result.options as any).configFile
```
Comment (johnnyreilly) [inline src/config.ts]: Is the `any` necessary?

**Comment by artem-malko** [comment]
@andrewbranch hi, how is it going with real life project?)

**Comment by andrewbranch** [comment]
Nothing new to add since https://github.com/TypeStrong/ts-loader/pull/935#pullrequestreview-247345616

**Comment by marty-wang** [comment]
what is holding off the merge?

**Comment by MLoughry** [comment]
It's using a feature that hasn't yet been published in a mainline release.

**Comment by artem-malko** [comment]
https://github.com/microsoft/TypeScript/pull/31432 is merged and published, as I understand)

**Comment by intojun** [comment]
I check this out everyday. Any news here?

**Comment by johnnyreilly** [comment]
Patience @intojun - this is open source, it's summertime and people are busy. Things will happen I expect.... But I can't say when 😃

**Comment by mc0** [comment]
TypeScript 3.6 stable was released yesterday and it appears the failing check was just due to a network error with installing a package. Beyond that passing, is there anything I can do to help this along?

**Comment by AndrewKirkovski** [comment]
CI is failing because of `invalid json response body at https://registry.yarnpkg.com/@uifabric%2Fstyling reason: Unexpected end of JSON input`, what's holding to recheck and merge?

**Comment by johnnyreilly** [comment]
I appreciate lots of people are waiting on this. It's probably worth doing some kind of status update to share where things are to my knowledge.

Whilst this is mergeable, to my knowledge @andrewbranch still had some work he planned to do; see comment here: https://github.com/TypeStrong/ts-loader/pull/935#issuecomment-510322126

I believe @andrewbranch maybe hoping to loop back round and take another look at this. There's no timeframe on this and I'm very grateful for all the help he provides as and when he's able to. Likewise @sheetalkamat. However, both of them have busy lives building TypeScript itself which I'm very grateful for. ♥️ 

@andrewbranch if you read this and there's a way to help others help you with this then that's amazing. If it's probably just easier for you to take a look when you get a moment; that's also great. Whatever works.

I understand people want this, but I'd encourage us all to be patient, kind and grateful.

**Comment by andrewbranch** [comment]
Hey all! I owe you an apology. I was confused on the state of this PR 😬

I had intended to try this PR out end-to-end while we were still waiting on 3.6 to land and while Sheetal was wrapping things up. Other issues (and my workload for TypeScript 3.7) are still making that difficult, but I totally missed the fact that Sheetal was done here. I was confused by the fact that the build is still red (which as @AndrewKirkovski pointed out, looks to be just an infrastructure hiccup) and by the fact that package.json is still referencing the nightly. I’ve just pushed an update fixing that. We probably could have merged this when 3.6 stable was released, and it was totally my fault for not paying close enough attention. I’m really sorry to everyone who was anxiously awaiting this feature.

I’ll also be updating the changelog shortly, then I think we can merge this.

In the meantime, has anyone else tried this PR in a real project? That’s certainly the single most helpful thing anyone can do with any large PR like this.

**Comment by kirill-konshin** [comment]
Does this PR address incremental build too?

**Comment by andrewbranch** [comment]
@kirill-konshin yes (notice that `.tsbuildinfo` is added to the .gitignore of the execution tests—that’s because the build info file is getting generated 🎉)

**Comment by kirill-konshin** [comment]
@andrewbranch yes, I just received your comment in incremental-related issue! Perfecto!

**Comment by johnnyreilly** [comment]
> I’ll also be updating the changelog shortly, then I think we can merge this.

Nice work @andrewbranch! Thanks so much!

If you're happy with this do feel free to merge and push out a release (just add an entry to the release to the https://github.com/TypeStrong/ts-loader/releases and the GitHub action should take care of the rest)

**Comment by paulvanbrenk** [comment]
Thanks @andrewbranch & @sheetalkamat

**Comment by andrewbranch** [comment]
This is in the process of shipping as v6.1.0. My current thinking is to get some validation that things are generally working, then remove the `projectReferences` configuration option as a v7.0 if things look good / after any bug fixes from early feedback. I think there should be no reason to toggle project references on or off based on loader config when tsconfig is a perfectly good source of truth, but removing the option will be a breaking change, and I want to be extra sure the functionality is solid before removing the opt-in nature of the feature.

**Comment by ericanderson** [comment]
Due to random bugs that can happen, leaving the failsafe so regular tsc can use build references but webpack doesn’t have to is good. Maybe 7 could default to on?

**Comment by andrewbranch** [comment]
I think you’d be better off using a different tsconfig for ts-loader that makes the intention more explicit. In general, I want to move ts-loader away from specialized behavior that contradicts the tsconfig it’s reading. Removing the option from ts-loader doesn’t remove the option for the user, it just shifts it onto TypeScript itself.

**Comment by sublimator** [comment]
I have a (unfortunately not open source, yet ...) project that is using `projectReferences: true`, and for some reason webpack (using webpack-cli) stalls and doesn't exit after what seems like a successful build. 

Not sure how to provide more info:
```
    "ts-loader": "^6.1.0",
    "webpack": "^4.39",
    "webpack-cli": "^3.3.8",
    "ts-loader": "^6.1.0",
```

I happen to be using TypeScript for the webpack config itself, which may or may not indicate something.

**Comment by sublimator** [comment]
I just cloned this repository and "manually" ran webpack on the `test/execution-tests/3.6.0_projectReferencesToBeBuilt/main.js` and experienced the same hang. Note that I'm on MacOS Mojave yet the same hang is evident on CircleCI running linux inside docker (circleci/node:10.16.2-buster-browsers to be exact)

**Comment by sublimator** [comment]
Running `yarn test` from the root of the cloned repo

```
  projectReferences
    1) should have the correct output
    ✓ should work with transpileOnly (822ms)


  1 passing (4s)
  1 failing

  1) projectReferences
       should have the correct output:

      Uncaught AssertionError [ERR_ASSERTION]: bundle.js is different between actual and expected
      + expected - actual

       exports.lib = {
           one: 1,
           two: 2,
           three: 3
      +    / I am adding this comment here by hand to ensure
      +    / Webpack is using the JS output for project references
       };
       /# sourceURL=webpack://./lib/index.ts?");
       /***/ })
       /******/ });
      
      at compareActualAndExpected (test/comparison-tests/create-and-execute-test.js:467:16)
      at /Users/nicholasdudfield/projects/ts-loader/test/comparison-tests/create-and-execute-test.js:333:13
      at Array.forEach (<anonymous>)
      at compareFiles (test/comparison-tests/create-and-execute-test.js:329:31)
      at Watching.handler (test/comparison-tests/create-and-execute-test.js:194:9)
      at compiler.hooks.done.callAsync (node_modules/webpack/lib/Watching.js:100:9)
      at AsyncSeriesHook.eval [as callAsync] (eval at create (node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:32:10), <anonymous>:6:1)
      at AsyncSeriesHook.lazyCompileHook (node_modules/webpack/node_modules/tapable/lib/Hook.js:154:20)
      at Watching._done (node_modules/webpack/lib/Watching.js:99:28)
      at compiler.emitRecords.err (node_modules/webpack/lib/Watching.js:73:19)
      at Compiler.emitRecords (node_modules/webpack/lib/Compiler.js:366:39)
      at compiler.emitAssets.err (node_modules/webpack/lib/Watching.js:54:20)
      at hooks.afterEmit.callAsync.err (node_modules/webpack/lib/Compiler.js:352:14)
      at AsyncSeriesHook.eval [as callAsync] (eval at create (node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:32:10), <anonymous>:6:1)
      at AsyncSeriesHook.lazyCompileHook (node_modules/webpack/node_modules/tapable/lib/Hook.js:154:20)
      at asyncLib.forEach.err (node_modules/webpack/lib/Compiler.js:349:27)
      at done (node_modules/neo-async/async.js:2854:11)
      at /Users/nicholasdudfield/projects/ts-loader/node_modules/neo-async/async.js:2805:7
      at /Users/nicholasdudfield/projects/ts-loader/node_modules/graceful-fs/graceful-fs.js:45:10
      at FSReqWrap.oncomplete (fs.js:141:20)
```

Likewise `yarn test` does not seem to exit and stalled after the above output

**Comment by johnnyreilly** [comment]
@sublimator to do anything useful to help we're going to need a minimum reproduction repo please. If you can make one, raise an issue and link back to this with it that would be amazing.

**Comment by sublimator** [comment]
@johnnyreilly 

Hey, will do! 

But as stated above, all you need to repro is clone this repo and run the tests, either with yarn test, or manually invoking webpack on the "fixture", and it hangs

**Comment by AndrewKirkovski** [comment]
I have identical hanging to @sublimator happening on Windows 10 machine

**Comment by maclockard** [comment]
Definitely still curious about what the expected interaction between ts-loader with `projectReferences: true` and fork-ts-checker-webpack-plugin. Will fork-ts need to add project reference support as well?

**Comment by johnnyreilly** [comment]
I suspect so, yes. If you'd like to raise a PR we'd appreciate it!

**Pull request merged**